src/plugins/fs/udf.c

```diff
@@ -0,0 +1,503 @@
+    g_autofree gchar *vid = NULL;
+
+    if (options->label) {
+        /* there are different limits for lvid and vid and we need to set both so lets just
+           truncate vid (GParted does the same thing) */
+        if (g_str_is_ascii (options->label) && strlen (options->label) > 30)
+            vid = g_strndup (options->label, 30);
+        else if (g_utf8_validate (options->label, -1, NULL) && g_utf8_strlen (options->label, -1) > 15)
+            vid = g_utf8_substring (options->label, 0, 15);
+        else
+            vid = g_strdup (options->label);
+
+        g_ptr_array_add (options_array, bd_extra_arg_new ("--lvid", options->label));
+        g_ptr_array_add (options_array, bd_extra_arg_new ("--vid", vid));
+    }
+
+    if (options->uuid)
+        g_ptr_array_add (options_array, bd_extra_arg_new ("-u", options->uuid));
+
+    if (extra) {
+        for (extra_p = extra; *extra_p; extra_p++)
+            g_ptr_array_add (options_array, bd_extra_arg_copy ((BDExtraArg *) *extra_p));
+    }
+
+    g_ptr_array_add (options_array, NULL);
+
+    return (BDExtraArg **) g_ptr_array_free (options_array, FALSE);
+}
+
+static gint get_blocksize (const gchar *device, GError **error) {
+    gint fd = -1;
+    gint blksize = 0;
+
+    fd = open (device, O_RDONLY);
+    if (fd < 0) {
+        g_set_error (error, BD_FS_ERROR, BD_FS_ERROR_FAIL,
+                    "Failed to open the device '%s' to get its block size", device);
+        return -1;
+    }
+
+    if (ioctl (fd, BLKSSZGET, &blksize) < 0) {
+        g_set_error (error, BD_FS_ERROR, BD_FS_ERROR_FAIL,
+                     "Failed to get block size of the device '%s'", device);
+        close (fd);
+        return -1;
+    }
+
+    close (fd);
+
+    return blksize;
+}
+
+/**
+ * bd_fs_udf_mkfs:
+ * @device: the device to create a new UDF fs on
+ * @media_type: (allow-none): specify the media type or %NULL for default ('hd')
+ * @revision: (allow-none): UDF revision to use or %NULL for default ('2.01')
+ * @block_size: block size in bytes or 0 for auto detection (device logical block size)
+ * @extra: (allow-none) (array zero-terminated=1): extra options for the creation (right now
+ *                                                 passed to the 'mkudffs' utility)
+ * @error: (out): place to store error (if any)
+ *
+ * Returns: whether a new UDF fs was successfully created on @device or not
+ *
+ * Tech category: %BD_FS_TECH_UDF-%BD_FS_TECH_MODE_MKFS
+ */
+gboolean bd_fs_udf_mkfs (const gchar *device, const gchar *media_type, gchar *revision, guint64 block_size, const BDExtraArg **extra, GError **error) {
+    const gchar *args[7] = {"mkudffs", "--utf8", NULL, NULL, NULL, device, NULL};
+    gint detected_bs = -1;
+    gboolean ret = FALSE;
+
+    if (!check_deps (&avail_deps, DEPS_MKUDFFS_MASK, deps, DEPS_LAST, &deps_check_lock, error))
+        return FALSE;
+
+    if (block_size != 0)
+        args[2] = g_strdup_printf ("--blocksize=%"G_GUINT64_FORMAT"", block_size);
```

Comment:
I appreciate your futureproof attitude for huge block sizes :snowman:

Reply:
Teoretical maximal block size for UDF is 32768. But kernel supports block size up to the page size, so on x86 just 4096 and maybe on itanium with 8k pages it could be 8192.